### PR TITLE
data/aws/vpc: Add an elasticloadbalancing endpoint

### PR DESCRIPTION
**Pull request by wking**
To [avoid][1]:

    The DNS provider failed to ensure the record: failed to get hosted zone for load balancer target "afdffb7c0550244d39ac35bdfd21fff0-1591348991.us-east-1.elb.amazonaws.com": failed to describe load balancers: RequestError: send request failed caused by: Post https://elasticloadbalancing.us-east-1.amazonaws.com/: dial tcp 54.239.29.168:443: i/o timeout

when running the DNS provider in a blackhole where it can't get out to the wider internet.  With the LB endpoint exposed inside the VPC, the operator will be able to reach the cloud provider without needing to go through the proxy.  I expect we will want the ingress operator to respect proxy settings anyway, but at least with this change we won't *need* that for blackholed private subnets.

`docs/user/aws/images/install_upi.dia` needs to be updated to reflect the new endpoint, but I couldn't figure that out on short notice and it seems safe to punt.  There's also an issue with [tagging][1]:

    The DNS provider failed to ensure the record: failed to find hosted zone for record: failed to get tagged resources: RequestError: send request failed caused by: Post https://tagging.us-east-1.amazonaws.com/: dial tcp 52.94.233.76:443: i/o timeout

but apparently no VPC endpoints for tagging.  But one step at a time ;).

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_release/5308/rehearse-5308-pull-ci-openshift-installer-master-e2e-aws-proxy/10/artifacts/e2e-aws-proxy/must-gather/namespaces/openshift-ingress-operator/ingress.operator.openshift.io/dnsrecords/default-wildcard.yaml

### CODE REVIEW

**Comment by wking** [comment]
Added the Dia edits.

File: data/data/aws/vpc/vpc.tf [hunk 1]
```diff
@@ -18,6 +18,17 @@ resource "aws_vpc" "new_vpc" {
   )
 }
 
+resource "aws_vpc_endpoint" "elasticloadbalancing" {
```
Comment (abhinavdahiya) [inline data/data/aws/vpc/vpc.tf]: So i don't think we should be creating these endpoints in IPI as they don't provide great value to the cluster.

If they do, let's start a general discussion about these elsewhere, github issue / enhacements.

File: upi/aws/cloudformation/01_vpc.yaml [hunk 2]
```diff
@@ -244,6 +244,29 @@ Resources:
       DestinationCidrBlock: 0.0.0.0/0
       NatGatewayId:
         Ref: NAT3
+  LoadBalancingEndpoint:
```
Comment (abhinavdahiya) [inline upi/aws/cloudformation/01_vpc.yaml]: Again this UPI templates are reference only. And I don't think we allow special env setup to handle these and not add these for all.

This is not required for openshift clusters and don't provide value.

**Comment by abhinavdahiya** [comment]
/hold

I think we should have a larger discussion on IPI and VPC endpoints.

and the change to UPI templates _which is for reference only_ and _bare minimum for installs_ doesn't seem to need these endpoints.

If the proxy test needs to create an environment, i would hope the test handle those separately.

**Comment by wking** [comment]
> I think we should have a larger discussion on IPI and VPC endpoints.

Sounds good.  Spun off into #2486, because it's installer-specific and not a cross-team issue.

> and the change to UPI templates which is for reference only and bare minimum for installs doesn't seem to need these endpoints.

By this reasoning we should drop [our existing S3 VPC endpoint][1], which is for saving customers money (per #745) and allows for registry access without going through a proxy even from blackholed subnets.  And we should be very clear about which environments we are targeting as a bare minimum (e.g. the templates would not be sufficient for blackholed private subnets).  But I'm fine circling back to this after we settle #2486.

[1]: https://github.com/openshift/installer/blob/499f53fcbc89fdddd5541f00a9c97578bb7dbc6d/upi/aws/cloudformation/01_vpc.yaml#L247-L269

**Comment by abhinavdahiya** [comment]
> By this reasoning we should drop our existing S3 VPC endpoint, which is for saving customers money (per #745) 

because you pull data from s3 apis, it is definitely saving our customers money. the elb is not.

>  And we should be very clear about which environments we are targeting as a bare minimum (e.g. the templates would not be sufficient for blackholed private subnets).

The default / getting ready / basic installation. blacked-holed subnets created by the template is not that. is the user skip the vpc and wants to do more advanced that an exercise for them.

**Comment by danehans** [comment]
@wking I don't believe this PR is required for proxy. The ingress operator calls the aws elb api, but that point it moot if we state https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#remove-dns-zones is required for proxy black-holed vpc setups. The k8s controller-manager calls the elb api when a svc of type `LoadBalancer` is created. The controller-manager did not properly support proxy prior to https://github.com/openshift/cluster-kube-controller-manager-operator/pull/285. cc: @ewolinetz

**Comment by abhinavdahiya** [comment]
I think we need a better consensus on addition to IPI setup https://github.com/openshift/installer/issues/2486

and it seems like it doesn't help the UPI templates too https://github.com/openshift/installer/pull/2485#issuecomment-540750534 and https://github.com/openshift/installer/issues/2486#issuecomment-540245518

So closing this for now. we can use the github issue to discuss this before opening a change.

/close

**Pull request closed without merging**